### ISSUE TITLE
Docs enhancement - add a copy button to code examples 

### ISSUE THREAD

**Issue by shannonbux** [gatsbyjs/gatsby#5030]
## Summary
Code examples in tutorials, READme files, and guides could use a copy button so people don't miss a single thing when they copy code.

## Motivation 
When we did a Gatsby workshop a while ago, many workshop attendees' computer screens didn't show backtics very clearly and they missed copying them a lot.

Another related issue brought up by @twhiteaker is that when copying code from Firefox to VS code (and Notepad ++), the code ended up having two carriage returns added to the top and bottom. This actually mattered for a markdown file, it appears.

## Acceptance criteria (by @marcysutton)

- As a keyboard user, I should be able to reach and operate a copy button for each code example and see a visible focus style.
- As a sighted user, I should see a graphic icon with adequate color contrast for each code example that is intuitive and clear that it will copy code to the clipboard. 
    - If the icon button is unclear through user testing, adding the text “Copy” or “Copy to clipboard” may help
- As an assistive technology user, when navigating through the copy buttons I should hear unique button names like `${filename}: copy code to clipboard` to know which file I’m copying to the clipboard.
- As a mouse user, clicking on the button should copy the adjacent code to the user’s clipboard in all major browsers.
- As a keyboard or screen reader user, focusing on the button and hitting the ENTER or SPACE keys should copy the adjacent code to the clipboard in all major browsers.
- As a sighted user, upon activating the button, I should see a visible “copied to clipboard!” toast notification.
- As a keyboard user, upon activating the copy button and toast notification, I should be able to hit the TAB key to reach a visible “Close” or “X” button with a visible focus style. When the toast notification is not visible, the button should be rendered inoperable.
- As a user with limited mobility or a cognitive impairment, I should have time to read the toast notification before it is dismissed by only closing it when activating the close button.
- As a sighted user, I should be able to dismiss the toast notification by clicking the close button or anywhere else on the screen. 
- As a screen reader user, I should hear an announcement of `${filename} copied to clipboard` after activating the button. 
    - This can be accomplished with a persistent, visually hidden DOM element with ARIA `role=”status”.
    - Add and announce a keyboard shortcut, like “ENTER” to close? (We can play with it)

Potential icon: 

![copy to clipboard icon](https://user-images.githubusercontent.com/1045233/59122850-ab6c6300-8910-11e9-9d73-1603296144a6.png)

Alternatively, there could be a button that just visually says "copy" without an icon.

**Comment by shannonbux** [comment]
@DSchau and @darpan I can do this! I'm assigning it to myself now. Thanks for doing the work, @DSchau !!!

**Comment by DSchau** [comment]
@shannonbux I didn't do the copy stuff yet! But you can definitely do the code title stuff if you're interested!

**Comment by shannonbux** [comment]
@fk can you add that reference you found to a copy button that you shared with @DSchau recently?

**Comment by DSchau** [comment]
This would be an _awesome_ improvement for users browsing gatsby documentation.

I'm sure @fk has some ideas regarding design, but let's build it out functionally first and then we can augment that existing functionality.

I think it's a fair assumption that users of this plugin will _also_ be users of `gatsby-remark-prismjs`, so I think we can build it as a `gatsby-remark-` plugin or built into `gatsby-remark-prismjs` itself. There are benefits to both approaches, so up for whatever whoever tackles this is leaning towards!

## Further Detail

- This should enable a `COPY` button (or svg icon!), that when clicked, will copy the text of the snippet that the copy button refers to
- We should use the [Clipboard API](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API)
    - I'm not _super_ concerned about a fallback here--although that's always welcome--so let's consider this a progressively enhanced feature
- The button probably needs some existing structure (e.g. HTML wrapper created with `gatsby-remark-prismjs`) so it might be a good idea to build it in at the plugin level, and enable this feature like so:

```js
// gatsby-config.js
plugins: [
  {
    resolve: `gatsby-transformer-remark`,
    options: {
      plugins: [
        {
          resolve: `gatsby-remark-prismjs`,
          options: {
            copy: true
          }
        }
      ]
    }
  }
]
```

_or_ an alternative approach is a standalone plugin that _depends_ on the structure created by `gatsby-remark-prismjs`. I think that's more error prone, so I'd lean towards the former, but both are reasonable approaches.

```js
// gatsby-config.js
plugins: [
  {
    resolve: `gatsby-transformer-remark`,
    options: {
      plugins: [
        `gatsby-remark-prismjs`,
        `gatsby-remark-copy-snippet`
      ]
    }
  }
]
```

there are pro's and con's to both approaches, so this should be a fun one to work on 💪

**Comment by DSchau** [comment]
@felipecaputo I can mark this as taken if you're interested! Let me know!

**Comment by felipecaputo** [comment]
That would be great. I'll take a look at the plugin tonight and get back to
you

**Comment by DSchau** [comment]
Done!

**Comment by Manoz** [comment]
Bootstrap's documentation is a good example I guess :)
![image](https://user-images.githubusercontent.com/1554207/46700644-ab9b4e00-cc1d-11e8-90d0-779615fcc95f.png)

**Comment by tanvibhakta** [comment]
Hey @felipecaputo Have you had a chance to look at the plug in yet?

**Comment by felipecaputo** [comment]
@tanvibhakta I was studying the code last week, but this week I couldn't make it because we were on a rush here at work. I believe I'll finish this till the end of this weekend!

**Comment by shannonbux** [comment]
@felipecaputo do you still have the bandwidth to tackle this issue? I've removed the "claimed" label. Let me know if you have a WIP PR and we can add that label back here!

**Comment by alexboffey** [comment]
@shannonbux I'd happy to pick this up! I'm new to Gatsby (and OS as a whole) as a contributor, but this seems like a good start 😄

**Comment by patricksimonian** [comment]
This is actually a user story at my work to enable this feature. @alexboffey have you started on this? I'd be interested in watching/contributing if I can to your project. If not I'd be looking into doing my own implementation at some point within the next few sprints here.

**Comment by DSchau** [comment]
@patricksimonian hey, awesome!

I've thought a bit more about this feature, and I think it _may_ sense to "fork" my repo [`gatsby-remark-code-titles`](https://github.com/dschau/gatsby-remark-code-titles), and add in copy functionality since it could then easily re-use that header structure that it creates.

So we could call this plugin `gatsby-remark-code-header` or something, and it could have a configurable title/copy button/etc. The high level API would look something like this:

```js
module.exports = {
  plugins: [
    {
      resolve: 'gatsby-remark-code-header',
      options: {
        copyButton: true, // default: true
        title: true // default: true
        // anything else we want
      }
    }
  ]
}

```

this way we can sorta side step any weirdness with other plugins implementing/changing the HTML structure, and can rather be in control of what we output, where, and even _if_ we should render a certain structure.

This _could_ also be structured as a plugin that takes other plugins (e.g. a plugin to generate a copy code button), but I think it makes sense to actually _deliver_ the functionality first and then we can look at whether optimizing makes sense!

**Comment by patricksimonian** [comment]
@DSchau wow sounds great. I'll take a look into your repo over the next coming weeks. I'm liking your thoughts on allowing for other plugins. I do agree that delivering functionality is probably best at first. At the very least, allowing for a className to be appended to the 'icon' element would allow for *some* flexibility. 

Like i said, I know this is a feature thats been requested at my work place. I'm not going to be able to work on it yet until it's apart of my Sprint. I'd anticipate that it will be within the next few weeks (latest start of December). Does that timeline work for you?

**Comment by DSchau** [comment]
@patricksimonian definitely, there's no required deadline or anything for this plugin, but as always... the sooner the better! This would be a big enhancement for not only our docs but a lot of other sites that use Gatsby for documentation!

**Comment by alexboffey** [comment]
Hey @patricksimonian 😄 
So far, just set up the codebase on my machine and been fiddling around. If I make any progress I will update this thread with a link to the topic branch.

**Comment by krvajal** [comment]
Hey, I got something working here (https://github.com/DSchau/gatsby-remark-code-titles/pull/1). Is this the right direction to go. I would love to get some feedback on it and try to get it working.

**Comment by DSchau** [comment]
@krvajal I'm currently at GraphQL Summit, but I'll try to take a look at this tonight or tomorrow. Thank you!

**Comment by iamskok** [comment]
Hi, guys! Is there any progress with this feature?

**Comment by patricksimonian** [comment]
Sorry i've been swamped at work, we ended up forgoing this feature as apart of our sprints in favor of other ones so I haven't had a chance to take a look

**Comment by DSchau** [comment]
@iamskok sure!

One note here: a PR was opened up on one of my repos that implements this feature.

I haven't had a ton of time to take a look, but feel free to take that as a base!

https://github.com/DSchau/gatsby-remark-code-titles/pull/1

**Comment by iamskok** [comment]
Thanks, @DSchau for pointing me to https://github.com/DSchau/gatsby-remark-code-titles it was very 
helpful for getting my head around markdown parsing.

I was thinking about various use cases of this plugin usage and came to the conclusion that it's better to create two separate plugins. Because not every blog requires copy functionality and not everyone 
requires code titles and probably it's even a more rare use case to have both at the same time.

While thinking about the plugin for achieving click to copy functionality I came to a point that we limit ourselves and buttons may be used for a wider range of actions (e.g. share, switch syntax highlighting theme, etc.). It's worth thinking about writing a more general purpose plugin,
which in future might be extending with other actions. 

Really liked your idea of using `:` followed by a key-value pair and suppose we should stick 
to this way of passing properties.

`js:title=example-file.js`

Though there are some problems with this implementation in `gatsby-remark-code-titles`. In your package, you are assuming that there are only 2 parameters (`language` and `title`) separated with `:`, but it may not always be the case in this implementation. When the array is split it takes
only first two parameters and removes the rest.  Also, `title` might not always be the first parameter. It seems logical to always keep `language` as a first parameter.
https://github.com/DSchau/gatsby-remark-code-titles/blob/78c0b9beedee5ad562cc20fc71e8fcdff36775b7/src/index.js#L6

`js:title=example-file.js:action=true:other-thing=foo:something-else=bar`

Please let me know your thoughts.

**Comment by iamskok** [comment]
Hey, guys! 
Please checkout [gatsby-remark-code-buttons](https://github.com/iamskok/gatsby-remark-code-buttons) and let me know what you think.

Can someone assign me to this issue, please?

**Comment by fk** [comment]
@iamskok Niiiiiice! 😍 🙏 

I'm the wrong one to do a "technical review", but I'm sure @DSchau & Co. will take a look ASAP!
I'll jump ahead and assign you to the issue now!

**Comment by iamskok** [comment]
Thanks @fk :) For some reason I didn't get an invite.

**Comment by iamskok** [comment]
@DSchau Did you have a chance to checkout  [gatsby-remark-code-buttons](https://github.com/iamskok/gatsby-remark-code-buttons)?

**Comment by DSchau** [comment]
@iamskok I'll take a look now! At a glance, it looks good--mind PRing it to our `www` directory (aka gatsbyjs.org)?

One note is that I'm not sure I like the opt-in behavior. I think it should be enabled, by default! opt-out perhaps makes more sense?

**Comment by iamskok** [comment]
@DSchau Thanks for checking it out!

Do you mind clarifying what opt-in/opt-out behavior are you talking about? 

 > mind PRing it to our www directory (aka gatsbyjs.org)?

Sure, no problem! I'm gonna do it once the code will be ready.

**Comment by DSchau** [comment]
@iamskok it _appears_ that you have to add a specific handler, e.g.

````
```js:action=copy
alert('click to copy 💾');
```
````

correct?

**Comment by iamskok** [comment]
Yes, this is how it's working right now.

**Comment by DSchau** [comment]
Right! So I think we'd generally want _all_ code snippets to have the functionality. I'd prefer something like:

````
```js:copy-false
var a = 'this can not be copied'
```
````

I think! I'm open to however you prefer building in this functionality, but I think generally most will want this opt-out rather than opt-in!

**Comment by iamskok** [comment]
> I think! I'm open to however you prefer building in this functionality, but I think generally most will want this opt-out rather than opt-in!

That sounds fair.  Though I'd prefer key-value pair options:

``````
```js:copy=false
var a = 'this can not be copied'
```
``````

 I'll update the code today. Once you will say that everything looks good I'll push it to `www` directory. 

> Are you thinking there'd be some other action that could be done? Additionally, you could just default action to copy and tweak behavior if it's something else.

I don't have plans on adding any other functionality at the moment, but I can see it being extendable to other feature.

**Comment by iamskok** [comment]
@DSchau added default opt-out behavior iamskok/gatsby-remark-code-buttons@8c20ac607fbc2f19108590b2d20315426785b07f

* What branch should I push it to?
* Do you want me to set it up on [gatsbyjs.org](https://gatsbyjs.org) as well? If so we have to come up with custom styles for buttons and/or tooltips first.

**Comment by iamskok** [comment]
Made some edits related to @DSchau [comments](https://github.com/DSchau/gatsby-remark-code-titles/pull/2#pullrequestreview-200297076) for [gatsby-remark-code-titles](https://github.com/DSchau/gatsby-remark-code-titles)

iamskok/gatsby-remark-code-buttons@52752d378a6fac15ca84091a7f4ccd07e156fe1d

**Comment by ffdead** [comment]
Hi guys, not sure where to report this but I just tried to use `gatsby-remark-code-buttons` in combination with `gatsby-remark-prismjs` and `gatsby-remark-component`. The latter component causes this error to be thrown in the console:
```
Warning: Invalid event handler property `onclick`. Did you mean `onClick`?
```

**Comment by iamskok** [comment]
@ffdead please open this issue on https://github.com/iamskok/gatsby-remark-code-buttons and I'll have a look at it.

**Comment by shannonbux** [comment]
We are waiting on getting the gatsbyjs.org site working with mdx before
tackling this issue!

**Comment by rlynn523** [comment]
Is this issue still available? If so, I would love to take it.

**Comment by iamskok** [comment]
@rlynn523 this issue was closed for inactivity, but I believe it's still a wanted feature. You can use [gatsby-remark-code-buttons](https://github.com/iamskok/gatsby-remark-code-buttons) plugin to add copy buttons to the docs.

**Comment by LekoArts** [comment]
@rlynn523 Hi! This issue is currently "on hold" so to speak until https://github.com/gatsbyjs/gatsby/pull/12514 is merged

**Comment by rlynn523** [comment]
@LekoArts is it possible to assign this issue to me once that merge has completed?

**Comment by iamskok** [comment]
Added toaster notification option for [gatsby-remark-code-buttons](https://github.com/iamskok/gatsby-remark-code-buttons).

![](https://media.giphy.com/media/hoHRea4IdkDBrsE4Bu/giphy.gif)

**Comment by shannonbux** [comment]
oh hoh @iamskok nice! @m-allanson @marcysutton any idea if this would be an easy-enough interim solution before we get mdx implemented in Gatsby docs?

**Comment by marcysutton** [comment]
We're super close to merging the MDX PR, and I would need to test `gatsby-remark-code-buttons` for accessibility before integrating it–so I think we should hold off for a short period of time. I also have a task to determine scope for all the MDX components we want on the .org site, so I'll be sure to account for this one and circle back here when I know more.

**Comment by iamskok** [comment]
@marcysutton I currently don't have the capacity to test and implement accessibility features, but I will be more than happy to merge these PRs in  [gatsby-remark-code-buttons](https://github.com/iamskok/gatsby-remark-code-buttons).

**Comment by ar-stackrox** [comment]
@iamskok Does gatsby-remark-code-buttons work with gatsby-mdx as of now? I'm encountering these issues.

![image](https://user-images.githubusercontent.com/28272783/58128875-6dabd280-7bcd-11e9-9461-425a005f332b.png)

**Comment by iamskok** [comment]
@ar-stackrox it was never tested with `gatsby-mdx `. Can you please share a repo where I can reproduce this error?

**Comment by ar-stackrox** [comment]
@iamskok I can't right now because the repo is private, but all I did was an npm install, and then add it to the gatsby-config / gatsby-remark-plugin like so - 
![image](https://user-images.githubusercontent.com/28272783/58130397-c29d1800-7bd0-11e9-8524-67fa893f4c36.png)

**Comment by DSchau** [comment]
@ar-stackrox if at all possible--I'd request that you please re-open the issue in the [package itself](https://github.com/iamskok/gatsby-remark-code-buttons).

Thanks @iamskok for chiming in and being so helpful--we sincerely appreciate it!

**Comment by jmolivas** [comment]
@iamskok I can confirm the same issue when using MDX as reported by @ar-stackrox

**Comment by iamskok** [comment]
If anyone can reproduce this bug, please share it [here](https://github.com/iamskok/gatsby-remark-code-buttons/issues/1)

**Comment by marcysutton** [comment]
This component was spec'd as part of https://github.com/gatsbyjs/gatsby/issues/14095 (currently needing to be prioritized amongst other work, so if anyone wants to pick this up that would be fine by us as long as it meets the requirements). 

Here are some success criteria for what we need for this component to be successful:

## Success criteria
- As a keyboard user, I should be able to reach and operate a copy button for each code example and see a visible focus style.
- As a sighted user, I should see a graphic icon with adequate color contrast for each code example that is intuitive and clear that it will copy code to the clipboard. 
    - If the icon button is unclear through user testing, adding the text “Copy” or “Copy to clipboard” may help
- As an assistive technology user, when navigating through the copy buttons I should hear unique button names like `${filename}: copy code to clipboard` to know which file I’m copying to the clipboard.
- As a mouse user, clicking on the button should copy the adjacent code to the user’s clipboard in all major browsers.
- As a keyboard or screen reader user, focusing on the button and hitting the ENTER or SPACE keys should copy the adjacent code to the clipboard in all major browsers.
- As a sighted user, upon activating the button, I should see a visible “copied to clipboard!” toast notification.
- As a keyboard user, upon activating the copy button and toast notification, I should be able to hit the TAB key to reach a visible “Close” or “X” button with a visible focus style. When the toast notification is not visible, the button should be rendered inoperable.
- As a user with limited mobility or a cognitive impairment, I should have time to read the toast notification before it is dismissed by only closing it when activating the close button.
- As a sighted user, I should be able to dismiss the toast notification by clicking the close button or anywhere else on the screen. 
- As a screen reader user, I should hear an announcement of `${filename} copied to clipboard` after activating the button. 
    - This can be accomplished with a persistent, visually hidden DOM element with ARIA `role=”status”.
    - Add and announce a keyboard shortcut, like “ENTER” to close? (We can play with it)

Potential icon: 

![copy to clipboard icon](https://user-images.githubusercontent.com/1045233/59122850-ab6c6300-8910-11e9-9d73-1603296144a6.png)

Alternatively, there could be a button that just visually says "copy" without an icon.

**Comment by iamskok** [comment]
@marcysutton thanks for providing all the requirements. I can continue working on this issue, but I will need some guidelines with reproducing and fixing MDX compatibility bug. Do you know anyone who will be able to help me with it?

**Comment by gillkyle** [comment]
Hey @iamskok, I'm not sure I can provide a ton of help on debugging remark plugins as sub-plugins of `gatsby-mdx`, however after reading through this issue I'm wondering if a remark plugin is still the best way to go. That direction was encouraged when MDX wasn't really an option, then this was held off until MDX was merged in (adding the potential for this to just be a component) like @marcysutton's comment mentions:

> This component was spec'd as part of #14095 (currently needing to be prioritized amongst other work...

This is a great use case for MDX, but maybe @DSchau has an opinion since it would probably require reworking packages like `gatsby-remark-code-titles` and `gatsby-remark-prismjs` (which are making those code blocks now) into a new component.

I think using the remark plugin is still a good possibility and I don't have the most context around this either.

**Comment by Blackglade** [comment]
This plugin seems to have conflicts with `gatsby-remark-prismjs`. I need to specify this plugin before the `gatsby-remark-prismjs` otherwise it won't render. Also the styling conflicts with remark as well. Any ETA on when this will be moved to an official plugin

**Comment by iamskok** [comment]
@Blackglade can you please provide more information on styling conflicts you are talking about?

Not sure about ETA, I’ll need help from someone who is familiar with MDX to make it happen.

**Comment by Blackglade** [comment]
@iamskok So I have the following plugins listed under `gatsby-transformer-remark` in the following order:

* `gatsby-remark-code-buttons`
* `gatsby-remark-autolink-headers`
* `gatsby-remark-prismjs`
* `gatsby-remark-images`

Anytime I include `gatsby-remark-code-buttons` after `prism-js`. The icon just doesn't appear. and digging through the html, it doesn't seem as if the divs were even rendered. As for the styling issue, ignoring the black icon of the clipboard (which can be changed), it's off center. I am using the prism-okaidia theme btw <img width="1326" alt="Screen Shot 2019-07-02 at 11 33 24 PM" src="https://user-images.githubusercontent.com/6363077/60561673-5e6a8980-9d22-11e9-9f5f-c5d5e7db5000.png">

**Comment by iamskok** [comment]
@Blackglade please try this order:

```
{
  resolve: `gatsby-transformer-remark`,
  options: {
    plugins: [
      `gatsby-remark-code-buttons`,
      `gatsby-remark-code-titles`,
      `gatsby-remark-autolink-headers`,
      `gatsby-remark-prismjs`,
      `gatsby-remark-images`
    ]
  }
}
```

The working example of using all of these plugins together can be found in [gatsby-dev-blog-starter](https://github.com/iamskok/gatsby-dev-blog-starter).

[gatsby-remark-code-buttons](https://github.com/iamskok/gatsby-remark-code-buttons) provides generic styles to make positioning and adding tooltips easier, but the developer will be responsible for tweaking the button styles to match a syntax highlighting theme.

**Comment by KyleAMathews** [comment]
👍 to doing this — just saw someone again at a workshop run into an error with a graphql query because they didn’t copy the closing backtick.
